src/Lean/Data/LParsec.lean

```diff
@@ -0,0 +1,138 @@
+
+/-
+Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Dany Fabian
+-/
+
+namespace Lean
+
+namespace LParsec
+
+inductive ParseResult (α : Type)
+| success (pos : String.Iterator) (res : α)
+| error (pos : String.Iterator) (err : String)
+deriving Repr
+
+open ParseResult
+
+def Parser (α : Type) : Type := String.Iterator → ParseResult α
```

Comment:
yeah, I don't quite know. I think we can probably express the JSON parsing using `LParsec`, but not the other way round. The `LParsec` stuff is a more classical backtracking parser.

I initially was looking at `Quickparse` but didn't feel comfortable changing it too much to not break things.

The `LParsec` parser can be fast if you use the combinators right but it certainly can do arbitrary backtracking, too, so I don't quite know.

I'd put it this way; I think the `LParsec` parser is certainly more complete than the `Quickparse` one.

Do you think we should add the missing combinators from `Quickparse` and get rid of it, refactoring JSON to use `LParsec` instead?

Also, I deliberately kept them separate to have exactly this conversation.

Reply:
The reason I originally wrote quickparse in this way is for performance.  It can do backtracking just fine, it's just that the json parser doesn't need it.

I think it would be a good idea to extract the generic Quickparse code from the JSON parser and add backtracking combinators.  There's certainly no need to have two of them.

If you decide to rewrite the JSON parser one day, you should do benchmarks to see if there's a performance difference.  (But I don't see any urgent need to rewrite it.)

Reply:
@gebner, my main question would be more around how do you think we can converge the two parser combinators (not the concrete xml or json parsers).

If I said rewrite the JSON parser, then I mean simply using different functions to achieve the same thing, but I'd really want to keep exactly the same function bodies.

When I was translating the XML spec, I just noticed, that I had to add so many combinators and could re-use so little from quickparse, that I wrote this from scratch. I think the XML and JSON parser have almost nothing in common apart from the first 40 lines.

Reply:
Yes, it looks like what they share is mostly the type, the monadic instance, fail, and anyChar/next, which in fact look identical, so replacing those parts of Quickparse with LParsec should not change behavior or performance.

Reply:
> I think the XML and JSON parser have almost nothing in common apart from the first 40 lines.

Honestly, I don't see any reason why they should.  I believe even the definition of whitespace is different.  (It's also more than 40 lines, there are quite a few combinators that just have different names like pstring/expect, or next/anyChar, ...)

But the first 40 lines are still boilerplate that should be shared.  There parsing library is also extremely bare bones at the moment.  For example adding user-friendly error messages (with the input string highlighted/underlined at the error position) would be useful to both.

Reply:
> my main question would be more around how do you think we can converge the two parser combinators

Just to answer this explicitly: you've already almost done it, you only need to do `s/Quickparse/Parser/g` and `s/next/anyChar/g`, etc.  I can't see any semantic difference here.

Though for the record: I dislike the `LParsec.Parser` name: 1) the `LParsec` has an unnecessary `L`, and 2) `Parser` clashes with `Parser`.

Reply:
haha, ok, I'll add the combinators and move the code away from JSON. One last thing; is quickparse still a good name in the presence of backtracking? Or put differently, what's "quick" in *quick*parse?

Reply:
just for the record, I took the names from the FParsec library. It felt like the closest cousin I could find around. Haskell's Parsec with being a monad transformer was quite a bit further away, also I didn't want to run into surprises with laziness.

Reply:
A long long time ago, the built-in (Lean) syntax parser used the `parsec` monad, and this parsec port was meant to provide error recovery, nice error messages, flexibility, etc., but not maximum speed.  The `quickparse` name then came about as a parser with a different tradeoff.  (Or maybe I'm confusing the order of events here, I don't remember anymore.)  In any case, the idea was that quickparse should compile to some reasonably straightforward C code: the generated functions should take a string/array and an index, and then load the character at that index, switch based on that character, etc.

I don't really care about the combinator names (though `pstring` doesn't really fit into the naming convention).  As for the monad name, another possibility is `Lean.Parsec` (and then put the combinators into the `Lean.Parsec` namespace).  Although I believe that the Haskell parsec is implemented in a very different way under the hood.

> Haskell's Parsec with being a monad transformer

I'm pretty sure the original parsec was around before monad transformers were even invented.

Reply:
Alright, I moved stuff to `Lean.Parsec` the last step will be refactor the JSON parser to use the `Parsec` one, adding whatever combinators are missing.

Reply:
finished moving everything. I think the PR is ready to merge...